compiler/rustc_middle/src/mir/consts.rs

```diff
@@ -319,15 +319,13 @@ impl<'tcx> Const<'tcx> {
     ) -> Result<ConstValue<'tcx>, ErrorHandled> {
         match self {
             Const::Ty(_, c) => {
-                // We want to consistently have a "clean" value for type system constants (i.e., no
-                // data hidden in the padding), so we always go through a valtree here.
-                match c.eval_valtree(tcx, param_env, span) {
-                    Ok((ty, val)) => Ok(tcx.valtree_to_const_val((ty, val))),
-                    Err(Either::Left(_bad_ty)) => Err(tcx
-                        .dcx()
-                        .delayed_bug("`mir::Const::eval` called on a non-valtree-compatible type")
-                        .into()),
-                    Err(Either::Right(e)) => Err(e),
+                if c.has_non_region_param() {
+                    return Err(ErrorHandled::TooGeneric(span));
+                }
+
+                match c.kind() {
+                    ConstKind::Value(ty, val) => Ok(tcx.valtree_to_const_val((ty, val))),
+                    _ => Err(tcx.dcx().delayed_bug("Unevaluated `ty::Const` in MIR body").into()),
```

Comment:
This is perhaps not quite right if we have a situation like `<u8 as Trait<impl Trait>>::ASSOC`. This would be a fully concrete rigid alias that is not normalized due to the opaque type. This is only possible to reach on nightly via unstable features that represent type system constants directly without an anon const.

I expect that long term we should just not encounter such a scenario by always normalizing type system constants in `Reveal::All` as CTFE unconditionally uses `Reveal::All` even during typeck *anyway*

Reply:
I think I mentioned leaving a FIXME below

Reply:
Nevermind, we always evaluate in `Reveal::All` so it should not be possible to encounter a fully concrete `ty::Const` that is also an alias as we would have simply evaluated it away.